stackslib/src/net/api/postblock_proposal.rs

```diff
@@ -541,8 +557,71 @@ impl NakamotoBlockProposal {
             builder.load_tenure_info(chainstate, &burn_dbconn, tenure_cause)?;
         let mut tenure_tx = builder.tenure_begin(&burn_dbconn, &mut miner_tenure_info)?;
 
+        let mut replay_txs_maybe: Option<VecDeque<StacksTransaction>> =
+            self.replay_txs.clone().map(|txs| txs.into());
+
         for (i, tx) in self.block.txs.iter().enumerate() {
             let tx_len = tx.tx_len();
+
+            // Check if tx is in the replay set, if required
+            if let Some(ref mut replay_txs) = replay_txs_maybe {
+                loop {
+                    let Some(replay_tx) = replay_txs.pop_front() else {
+                        // During transaction replay, we expect that the block only
+                        // contains transactions from the replay set. Thus, if we're here,
+                        // the block contains a transaction that is not in the replay set,
+                        // and we should reject the block.
+                        return Err(BlockValidateRejectReason {
+                            reason_code: ValidateRejectCode::InvalidTransactionReplay,
+                            reason: "Transaction is not in the replay set".into(),
+                        });
+                    };
+                    if replay_tx.txid() == tx.txid() {
+                        break;
+                    }
+                    let tx_result = builder.try_mine_tx_with_len(
+                        &mut tenure_tx,
+                        &replay_tx,
+                        replay_tx.tx_len(),
+                        &BlockLimitFunction::NO_LIMIT_HIT,
+                        ASTRules::PrecheckSize,
+                        None,
+                    );
+                    match tx_result {
+                        TransactionResult::ProcessingError(e) => {
+                            // The tx wasn't able to be mined. Check `TransactionError`, to
+                            // see if we should error or allow the tx to be dropped from the replay set.
+
+                            // TODO: handle TransactionError cases
+                            match e.error {
+                                ChainError::BlockCostExceeded => {
```

Comment:
I thought that this came through in a `TransactionSkipped`?

Reply:
Yep you're right - but also in fact, `BlockCostExceeded` is the wrong error code. Thanks for catching this and forcing me to dig into this a bit more. I've updated the logic here and also added a test for this case.